odoo/tools/pdf.py

```diff
@@ -19,3 +50,258 @@ def merge_pdf(pdf_data):
     merged_pdf = _buffer.getvalue()
     _buffer.close()
     return merged_pdf
+
+
+class OdooPdfFileWriter(PdfFileWriter):
+
+    def __init__(self, *args, **kwargs):
+        """
+        Override of the init to initialise additional variables.
+        :param pdf_content: if given, will initialise the reader with the pdf content.
+        """
+        super().__init__(*args, **kwargs)
+        self._is_pdfa = False
+        self._attachments_data = []
+        self._metadata = None
+
+    def cloneReaderDocumentRoot(self, reader):
+        """
+        Override from the base class to set the header properly, as well as give an id to the pdf.
+        The base class would always write PDF-1.3 without the additional bytes required by PDF/A, and always
+        remove the id.
+        """
+        super().cloneReaderDocumentRoot(reader)
+        # Set the PDF version to 1.7 (as PDF/A-3 are in version 1.7) and append some more bytes required by PDF/A
+        self._header = "%PDF-1.7\n%����".encode()
+        # Add a document ID to the trailer
+        pdf_id = ByteStringObject(md5(reader.stream.getvalue()).digest())
+        self._ID = ArrayObject((pdf_id, pdf_id))
```

Comment:
This requires a lot more explanations. Why md5, where does `_ID` come from, why does it insert the same data twice, ...

Reply:
From [Vera pdf validator](https://github.com/veraPDF/veraPDF-validation-profiles/wiki/PDFA-Parts-2-and-3-rules#rule-613-1)
> The first string is a permanent identifier based on the contents of the file at the time it was originally created and does not change when the file is incrementally updated. The second string is a changing identifier based on the file's contents at the time it was last updated. When a file is first written, both identifiers are set to the same value.

As for why md5: I based my code on the way Pypdf2 is creating this id. But it only does it when encrypting the pdf (as in the past, it was the only time where it was needed). But now with PDF/A, we need this information at all times.

Note: I have removed the manual creation of /ID to reuse the one from the incoming stream, so this code no longer exists.